### PR TITLE
Specified refactoring of facet related APIs

### PR DESCRIPTION
**Pull request by kore**
Specified refactoring of facet related APIs

### CODE REVIEW

File: doc/specifications/proposed/facets.md [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+additional queries) while this might be harder for the `$global` and
+`$filter` properties already existing in the `FacetBuilder`.
+
+### Facet Value Storage
+
+We should only store the actual IDs in the search backends, not the
+identifier. If a criterion is missing to query that ID we should just
+implement this (`SectionIdCriterion`). The queries are supposed to be
+more performant this way and it spares us re-indexing of the search
+index.
```
Comment (andrerom) [inline doc/specifications/proposed/facets.md]: .. if identifiers change

**Comment by andrerom** [comment]
Some topics that could need some further clarification:
- some standardization on facet sub types, in case of User this is `$type`
- labels, missing or covered by `Facet Return Values`?

**Comment by kore** [comment]
What do you mean by "labels"? We have the `$name` in there which could (should?) be used as a label for the facet in the frontend. Do you see labeling required for anything else?

Regarding `$type` vs. `$target`: I am all for cleaning things like this up, but seems a bit like an unnecessary BC break. On the other hand: We can deprecate one property and use interceptors to provide a BC. The `$target` is used in the criteria because of the generic constructor. `$type` seems like better naming to me, but we should probably go with the more established `$target` then. Shall I check through the `FacetBuilder` sub types to see which could use modification?

**Comment by andrerom** [comment]
Afaik `name` would be the user defined label for each facet, what I was getting at is to find away to avoid users having to load each and every object present on the facet entries to generate labels for those.  However this is clearly not SPI logic, but it might make sense for API to do this work on behalf of the user in a optimal way.

File: doc/specifications/proposed/facets.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+- RangeFacet
+- ListFacet
+
+Proposed Changes To The API
+---------------------------
+
+### Unsupported Facets
+
+If a storage engine does not support facets it should just ignore them
+and not throw a exception. The frontend code should expect that less
```
Comment (andrerom) [inline doc/specifications/proposed/facets.md]: Note: Done as of https://jira.ez.no/browse/EZP-26465 _(v1.10)_

File: doc/specifications/proposed/facets.md [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+instance before executing the query and make the search engine return this ID
+together with the facet return value.
+
+On way to generate such an ID could be as simple as:
+
+    foreach ($query->facets as $facetBuilder) {
+        $facetBuilderMap[md5(serialize($facetBuilder))] = $facetBuilder;
+    }
+
+Instead of `serialize()` we might want to use `spl_object_hash()` or something
```
Comment (andrerom) [inline doc/specifications/proposed/facets.md]: Note: Done in https://github.com/ezsystems/ezplatform-solr-search-engine/commit/d2d4fcb04600e71940598535285366c8e2ce944c using `spl_object_hash()`.

File: doc/specifications/proposed/facets.md [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+additional queries) while this might be harder for the `$global` and
+`$filter` properties already existing in the `FacetBuilder`.
+
+### Facet Value Storage
+
+We should only store the actual IDs in the search backends, not the
+identifier. If a criterion is missing to query that ID we should just
+implement this (`SectionIdCriterion`). The queries are supposed to be
+more performant this way and it spares us re-indexing of the search
+index if identifiers change.
```
Comment (andrerom) [inline doc/specifications/proposed/facets.md]: Note: Updated api doc to relfect in https://github.com/ezsystems/ezpublish-kernel/commit/13b710d9550677d7b803b4702bd9847c9879a0e7 _(v1.10, implementation in Solr was already doing this)_

**Pull request closed without merging**